### PR TITLE
Feature: Add/Edit conversation labels

### LINKED ISSUE #246: As an agent, I would like to see the conversations tagged to a particular label
- [ ] Clicking on labels should display all conversations labelled under that particular label

### PR DESCRIPTION
**Pull request by nithindavid**
Fixes #246 

## Preview
Opera Snapshot_2020-02-11_190144_localhost
Screenshot 2020-02-11 at 7 02 10 PM

### CODE REVIEW

**Comment by sojan-official** [comment]
@nithindavid  This PR looks great ❤️ . few thoughts based on the first look.

### Primary screen
- [ ]  Let's reevaluate the copy 'close conversation details'.  it seems to overflow and break the layout on many screens 
- [ ]  the box-shadow around conversation name and avatar made it seem like a button and little out of place. we can probably use a border-bottom to separate the information 
- [ ]   the add labels button looks out of place when there are no labels. maybe we can remove the copy and just show the empty text area instead.  
---
Chatwoot

### few other things 
----

- [ ] inconsistent copy, the drop-down says 'add this as a new tag' while we call it labels everywhere else 
Screenshot 2020-02-12 at 12 59 13 PM

- [ ] the `please enter to remove` and `test | selected` doesn't make sense since I don't have any indications inside the textarea 
Screenshot 2020-02-12 at 1 01 19 PM

- [ ]  we might be able to reduce the bottom margin a bit and accommodate more tags into the cloud
  
Screenshot 2020-02-12 at 1 01 59 PM

File: app/javascript/dashboard/routes/dashboard/conversation/ConversationLabels.vue [hunk 1]
```diff
@@ -38,25 +67,75 @@ export default {
...
+      labelUiFlags: 'conversationLabels/getUIFlags',
     }),
   },
   watch: {
     conversationId(newConversationId, prevConversationId) {
       if (newConversationId && newConversationId !== prevConversationId) {
-        this.$store.dispatch('conversationLabels/get', newConversationId);
+        this.$store
+          .dispatch('conversationLabels/get', newConversationId)
+          .then(() => {
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/conversation/ConversationLabels.vue]: Please add a catch for this statement, also I would prefer to use async await

File: app/javascript/dashboard/routes/dashboard/conversation/ConversationLabels.vue [hunk 2]
```diff
@@ -38,25 +67,75 @@ export default {
...
+            this.selectedLabels = [...this.savedLabels];
+          });
       }
     },
   },
   mounted() {
-    this.$store.dispatch('conversationLabels/get', this.conversationId);
+    this.$store
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/conversation/ConversationLabels.vue]: This seems to be repeated above, please create a function

File: app/javascript/dashboard/routes/dashboard/conversation/ConversationLabels.vue [hunk 3]
```diff
@@ -38,25 +67,75 @@ export default {
...
+        this.selectedLabels = [...this.savedLabels];
+      });
+  },
+  methods: {
+    addLabel(label) {
+      this.selectedLabels = [...this.selectedLabels, label];
+    },
+    asyncSearchLabels(query) {
+      this.isSearching = true;
+      ConversationAPI(query).then(response => {
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/conversation/ConversationLabels.vue]: no catch case

File: app/javascript/dashboard/routes/dashboard/conversation/ConversationLabels.vue [hunk 4]
```diff
@@ -38,25 +67,75 @@ export default {
...
+        this.countries = response;
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/conversation/ConversationLabels.vue]: Why is it labelled `countries`?

File: app/javascript/dashboard/i18n/locale/en/contact.json [hunk 5]
```diff
@@ -9,8 +10,9 @@
       "TITLE": "Previous Conversations"
     },
     "LABELS": {
-      "NO_RECORDS_FOUND": "There are no labels associated to this conversation.",
-      "TITLE": "Labels"
+      "TITLE": "Conversation Labels",
+      "UPDATE_BUTTON": "Update Labels",
+      "UPDATE_ERROR": "Couldn't update labels, try Again 🧐"
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/contact.json]: Please remove the emoji and it should be "try again"

File: app/javascript/dashboard/routes/dashboard/conversation/ConversationLabels.vue [hunk 6]
```diff
@@ -1,21 +1,42 @@
...
+    :class="hasEditedClass"
+  >
+    <div v-if="!conversationUiFlags.isFetching" class="wrap">
+      <div class="contact-conversation--list">
+        <label class="select-tags">
+          {{ $t('CONTACT_PANEL.LABELS.TITLE') }}
+          <multiselect
+            v-model="selectedLabels"
+            :options="savedLabels"
+            tag-placeholder="Add new label"
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/conversation/ConversationLabels.vue]: Please make this an i18n variable.

File: app/javascript/dashboard/routes/dashboard/conversation/ConversationLabels.vue [hunk 7]
```diff
@@ -1,21 +1,42 @@
...
+            placeholder="Search or add a label"
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/conversation/ConversationLabels.vue]: Please make this an i18n variable.

File: app/javascript/dashboard/components/widgets/Thumbnail.vue [hunk 8]
```diff
@@ -23,7 +23,7 @@
       src="~dashboard/assets/images/fb-badge.png"
     />
     <div
-      v-else-if="badge === 'Channel::WebWidget' && status === 'online'"
+      v-else-if="status === 'online'"
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/widgets/Thumbnail.vue]: Why was this removed?

File: app/javascript/dashboard/components/widgets/conversation/ConversationHeader.vue [hunk 9]
```diff
@@ -93,7 +93,7 @@ export default {
       ];
     },
     viewProfileButtonLabel() {
-      return `${this.isContactPanelOpen ? 'Hide' : 'View'}  Profile`;
+      return `${this.isContactPanelOpen ? 'Close' : 'More'} details`;
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/widgets/conversation/ConversationHeader.vue]: I guess we should move this also to i18n.

File: app/javascript/dashboard/store/modules/specs/conversationLabels/actions.spec.js [hunk 10]
```diff
@@ -32,4 +32,55 @@ describe('#actions', () => {
...
+    it('updates correct actions if API is success', async () => {
+      axios.post.mockResolvedValue({
+        data: { payload: { conversationId: '1', labels: ['on-hold'] } },
+      });
+      await actions.update(
+        { commit },
+        { conversationId: '1', labels: ['on-hold'] }
+      );
+
+      // commit(types.default.SET_CONVERSATION_LABELS, {
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/specs/conversationLabels/actions.spec.js]: Please remove these lines.

File: app/javascript/dashboard/components/widgets/Thumbnail.vue (hunk 8 above)
Comment (nithindavid) [reply app/javascript/dashboard/components/widgets/Thumbnail.vue]: Seems like polluting the component, Added the same case through params to the Thumbnail component .

Comment (pranavrajs) [reply app/javascript/dashboard/components/widgets/Thumbnail.vue]: The same is being used in a couple of places, repeating the logic doesn't seem good to me. I guess it is easier if the component can decide when to show status and when to show the badges
